### PR TITLE
TypeScript support

### PR DESCRIPTION
**Pull request by N2D4**
This adds the TypeScript compiler to the toolchain.

Our TypeScript support is powered by three tools:

- TypeScript itself, which we use to type-check our code.
- Babel, which we use to transpile TypeScript code into normal JavaScript code and to insert polyfills for older browsers
- Webpack, which takes the numerous ES6 modules and wraps them all into one big, compressed file

Any file that ends with `.js` will function exactly like before (except that they can now use ES6 imports/exports and other new features). However, files that end on `.ts` will have stricter typechecking enabled, and also a basic ESLint configuration to avoid common pitfalls. There are probably TypeScript and ESLint plug-ins available for your favorite text editor.

TypeScript does not particularly like global variables, so since we now have Webpack I split up the larger files into small ES6 modules which we can re-use. The source code is now in `src`.

Since we want users to be able to clone & immediately use the repo, the `dist` folder is not excluded in `.gitignore` like we usually would. There is a GitHub action to check that the `dist` folder is up-to-date. (Run `npm run build-prod` to update it)

`sdfv.html` (unlike `sdfv.css` and `sdfv.js`) was previously not part of the webclient, but part of the `diode` folder in the parent repo. As it fits in with this repo (and all its dependencies are here), I've moved it here, but I'm not sure what the implications of that are.

Since we now use Webpack, we can now use npm packages instead of having to include every single dependency as a `<script>` in our HTML. As a first step, I've done this with Dagre and MathJS - the jQuery plug-ins we use aren't available on npm, so I haven't changed anything about them.

This means that essentially everyone who still had a working copy of the webclient is going to get merge issues. Fortunately, since most of the code was just a lot of moving, it shouldn't be too hard to fix them. Let me know if you're getting a conflict.

Getting started with the new toolchain:

```
npm install
npm run serve
```

to launch the live-reload server (or `npm run watch` to only compile files)

### CODE REVIEW

File: src/utils/create_element.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/create_element.ts:1]:
```suggestion
// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: src/utils/deep_copy.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/deep_copy.ts:1]: it's quite weird to put every method in its own .ts file. Could you colocate createelement/deepcopy/equals/lerp as one utils.ts file?

File: src/utils/sdfg/check_and_redirect_edge.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/sdfg/check_and_redirect_edge.js:1]:
```suggestion
// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: src/utils/sdfg/check_and_redirect_edge.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+export function check_and_redirect_edge(edge, drawn_nodes, sdfg_state) {
```
Comment (tbennun) [inline src/utils/sdfg/check_and_redirect_edge.js:3]: same here for sdfg utility functions. It's a bit too many files that have the same purpose.

File: src/utils/sdfg/find_exit_for_entry.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/sdfg/find_exit_for_entry.js:1]:
```suggestion
// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: src/utils/sdfg/find_graph_element_by_uuid.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/sdfg/find_graph_element_by_uuid.js:1]:
```suggestion
// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: src/utils/sdfg/recursively_find_graph.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/sdfg/recursively_find_graph.js:1]: this could also go into some sdfg utils (see above comment)

File: src/utils/sdfg/sdfg_range_elem_to_string.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline src/utils/sdfg/sdfg_range_elem_to_string.js:1]: This file should be merged together with sdfg_property_to_string.js (maybe call them to_string.js or display.js?)

File: src/utils/sdfg/traverse_sdfg_scopes.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+
+/**
+ * Receives a callback that accepts (node, parent graph) and returns a value.
+ * This function is invoked recursively per scope (including scope nodes), unless the return
+ * value is false, upon which the sub-scope will not be visited.
+ * The function also accepts an optional post-subscope callback (same signature as `func`).
+ **/
+export function traverse_sdfg_scopes(sdfg, func, post_subscope_func = null) {
```
Comment (tbennun) [inline src/utils/sdfg/traverse_sdfg_scopes.js:9]: This file can go with memlet_tree and be called traversal.js

**Review by tbennun — CHANGES_REQUESTED** [review]
Thanks for doing this! There are a few extra files in dist/ that I'd be happy if we removed, plus some utility files can be merged together.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    "private": true,
+    "scripts": {
+        "clean": "rimraf dist",
+        "type-check": "tsc --noEmit",
+        "build-dev": "npm run clean && webpack --mode=development",
+        "watch": "npm run clean && webpack --watch --mode=development",
+        "serve": "npm run clean && webpack serve --mode=development",
+        "build-prod": "npm run clean && webpack --mode=production"
+    },
+    "author": "",
```
[the remark this answers is not in the document]
Comment (tbennun) [reply package.json:14]: If we really cannot remove this field (only if we can't), we can use "Scalable Parallel Computing Laboratory @ ETH Zurich" here

File: package.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    "license": "SEE LICENSE IN LICENSE",
```
[the remark this answers is not in the document]
Comment (tbennun) [reply package.json:15]: If we really cannot remove this field (only if we can't), we can use "New BSD, see LICENSE file" here

**Review by phschaad — CHANGES_REQUESTED** [review]
Thank you very much for this conversion, very nice work!

One small comment in addition to what has already been mentioned.

There is, however, a bigger point that needs to be discussed as well. The webclient forms an integral component of our [VSCode Extension](https://github.com/spcl/dace-vscode). The transition to a bundeled solution now brings two main issues with it:
1. We use the webclient's `renderer` variable (defined in `sdfv.js`) to interact with the renderer, which is now not defined outside its scope anymore. Can this somehow be made explicitly available externally?
3. We override certain webclient functions with VSCode specific behavior, effectively monkey patching certain webclient functions. This, too, is now not possible anymore. If we cannot find a way to monkey patch these functions from an external source, we will need to introduce a number of VSCode specific functions into the webclient explicitly.

I would also like to hear @tbennun's oppinion on these two issues if possible.

File: dist/626.js.map [hunk 12]
```diff
@@ -0,0 +1 @@
...           return chartData;\n        }).linkMouse(this.button_subwindow).changeGraphOptions(x => {\n            x.options.title.text = \"Cache misses (mean over all runs)\";\n            x.options.title.display = false;\n            x.options.legend = {\n                position: 'top',\n                display: true\n            };\n        }));\n\n        \n        databinding[\"PercentBandwidth\"] = all_mem_analyses;\n        databinding[\"Title\"] = new DataBlock({ fontsize: 32, text: \"Memory performance\", color: \"black\", align: \"center\" }, \"Text\");\n        databinding['Graph'] = all_mem_analyses;\n        databinding['Bandwidth'] = all_mem_analyses;\n\n        layout.setDataBinding(databinding);\n\n\n        this.button_subwindow.setLayout(layout);\n\n        this.setOnEnterHover(p => { this.color = \"#FF0000\"; this.button_subwindow_state = 'open'; })\n        this.setOnLeaveHover(p => { this.color = \"orange\"; if (!this.is_locked_open) this.button_subwindow_state = 'collapsed'; })\n        this.setOnClick((p, mb) => { this.is_locked_open = !this.is_locked_open; });\n\n        this.setDefaultDblClick();\n    }\n\n\n\n}\n\nexport { MemoryButton };"],"sourceRoot":""}
```
[the remark this answers is not in the document]
Comment (N2D4) [reply dist/626.js.map:1]: They're for lazy-loading scripts. JavaScript does that by calling `import(...)` at runtime, and there's some DIODE code which makes use of that. Since Webpack bundles scripts together, if `import(...)` is found it creates lazy-loadable bundles containing all the lazily-loaded dependencies.

Removing DIODE from the codebase gets rid of them.

**Comment by N2D4** [comment]
Thanks for your comments everyone!

@phschaad Fortunately the fix to those problems is pretty simple. The generated code still lives in the same context as before, it just doesn't make functions (and classes) available in the global namespace anymore. But that can be done manually:
```js
const renderer = ...;
globalThis.globallyAccessibleRenderer = renderer;  // `globallyAccessibleRenderer` is now in the global namespace
```
We can do the same for monkeypatching, but we'll need to specify all patchable functions in the webclient to make them accessible from global scope. Is [this](https://github.com/spcl/dace-vscode/blob/653b18c74c33f3508dba5b24107d89906b35d120/media/components/sdfv/sdfv_overrides.js) all the monkeypatching you do? If so, I'll add some logic so all those can be patched. Would you be okay with having to change a line in the webclient every time the list changes?

**Comment by phschaad** [comment]
This approach works for me, thank you for the suggestion! This list should be fairly consistent, so this works.

**Comment by N2D4** [comment]
@tbennun @phschaad Ready for another round!

**Comment by N2D4** [comment]
> @tbennun @phschaad Ready for another round of reviews!

I forgot to mention: I renamed the global variables to `daceRenderer` (for the renderer) and `daceUIHandlers.on_<method_name>` (for the monkey-patchable UI functions).

File: src/diode-ui/diode.js [hunk 13]
```diff
@@ -7,6 +7,7 @@ import { parse_sdfg, stringify_sdfg } from "../utils/sdfg/json_serializer";
 import * as DiodeTables from "./table";
 import { DIODE_Settings } from "./diode_settings";
 import { DIODE_Project } from "./diode_project";
+import { htmlSanitize } from "../utils/sanitization";
```
Comment (tbennun) [inline src/diode-ui/diode.js:10]: why is this necessary?

File: src/renderer/renderer.js [hunk 14]
```diff
@@ -60,12 +60,11 @@ export class SDFGRenderer {
         this.selected_elements = [];
 
         // Overlay fields
-        try {
-            this.overlay_manager = new OverlayManager(this);
-        } catch (ex) {
-            console.error("Error initializing overlay manager!", ex);
-            this.overlay_manager = null;
-        }
+		try {
```
Comment (tbennun) [inline src/renderer/renderer.js:63]: something changed this from spaces to tabs. Please format

File: src/sdfv.js [hunk 15]
```diff
@@ -390,86 +349,11 @@ function find_graph_element(graph, type, sdfg_id, state_id = -1, el_id = -1) {
...
-
-    sidebar_show();
-}
-
 function mouse_event(evtype, event, mousepos, elements, renderer,
     selected_elements, ends_drag) {
     if ((evtype === 'click' && !ends_drag) || evtype === 'dblclick') {
         if (renderer.menu)
-            renderer.menu.destroy();
+        renderer.menu.destroy();
```
Comment (tbennun) [inline src/sdfv.js]: strange indentation

**Review by tbennun — CHANGES_REQUESTED** [review]
Great work! Only minor comments this time around. Once fixed, and @phschaad approves the new `globalThis`, I will approve as well.

**Comment by phschaad** [comment]
Very nice work, thank you! The changes to `globalThis` with `daceRenderer` and the UI handlers look good and it seems like this works. Only one more thing that's tripping me up, which I unfortunately realized only now that I'm able to get that far thanks to your fixes: VSCode calls `init_sdfv`, and the utility functions `parse_sdfg`, and `stringify_sdfg` as well. It also creates a `new SDFG` object at one point. Is this something that can be facilitated easily? The utility functions could easily be duplicated out into VSCode if necessary, but `init_sdfv` and the `SDFG` class remain inaccessible.

**Comment by N2D4** [comment]
> Very nice work, thank you! The changes to `globalThis` with `daceRenderer` and the UI handlers look good and it seems like this works. Only one more thing that's tripping me up, which I unfortunately realized only now that I'm able to get that far thanks to your fixes: VSCode calls `init_sdfv`, and the utility functions `parse_sdfg`, and `stringify_sdfg` as well. It also creates a `new SDFG` object at one point. Is this something that can be facilitated easily? The utility functions could easily be duplicated out into VSCode if necessary, but `init_sdfv` and the `SDFG` class remain inaccessible.

Done, now [there is](https://github.com/spcl/dace-webclient/pull/29/commits/0fb11fc42420d05364a1f20cab71755fc402a386#diff-7a5091fdd5afee57eaea4d41b8ff3915fbe73c38bfd3374c16248eaa5f8c6b4cR21) `daceInitSDFV`, `daceParseSDFG`, `daceStringifySDFG` and `daceSDFGElements.SDFG`. If you need more, let me know!

By the way, would you prefer `dace.abc` or is `daceABC` okay?

**Review by phschaad — APPROVED** [review]
These changes work perfectly from the VSCode side of things, thank you very much for the nice work!

The naming is fine like this in my eyes :-)

**Review by tbennun — APPROVED** [review]
